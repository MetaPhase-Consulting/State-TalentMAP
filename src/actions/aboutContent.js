import { toastSuccess, toastError } from './toast';
import api from '../api';

export function aboutContentHasErrored(bool) {
  return {
    type: 'ABOUT_CONTENT_HAS_ERRORED',
    hasErrored: bool,
  };
}

export function aboutContentIsLoading(bool) {
  return {
    type: 'ABOUT_CONTENT_IS_LOADING',
    isLoading: bool,
  };
}

export function aboutContentFetchDataSuccess(data) {
  return {
    type: 'ABOUT_CONTENT_SUCCESS',
    data,
  };
}

export function aboutContentPatchHasErrored(bool) {
  return {
    type: 'ABOUT_CONTENT_PATCH_HAS_ERRORED',
    hasErrored: bool,
  };
}

export function aboutContentPatchIsLoading(bool) {
  return {
    type: 'ABOUT_CONTENT_PATCH_IS_LOADING',
    isLoading: bool,
  };
}

export function aboutContentPatchSuccess(success) {
  return {
    type: 'ABOUT_CONTENT_PATCH_SUCCESS',
    success,
  };
}

export function aboutContentFetchData() {
  return (dispatch) => {
    dispatch(aboutContentIsLoading(true));
    api().get('/position/1/')
      .then(() => {
        dispatch(aboutContentHasErrored(false));
        dispatch(aboutContentIsLoading(false));
        dispatch(aboutContentFetchDataSuccess(ABOUT_CONTENT)); // eslint-disable-line
      })
      .catch(() => {
        // TODO update
        dispatch(aboutContentHasErrored(true));
        dispatch(aboutContentIsLoading(false));
      });
  };
}

export function aboutContentPatchData(data) {
  return (dispatch) => {
    dispatch(aboutContentPatchIsLoading(false));
    api().get('/position/1/', { data })
      .then(() => {
        dispatch(aboutContentPatchHasErrored(false));
        dispatch(aboutContentPatchIsLoading(false));
        dispatch(aboutContentPatchSuccess(true)); // eslint-disable-line
        dispatch(aboutContentFetchDataSuccess(data)); // eslint-disable-line
        dispatch(toastSuccess('You may now press the cancel button or leave this page.', 'Update successful'));
      })
      .catch(() => {
        // TODO update
        dispatch(aboutContentPatchHasErrored(true));
        dispatch(aboutContentPatchIsLoading(false));
        dispatch(aboutContentPatchSuccess(false)); // eslint-disable-line
        dispatch(toastError('Update unsuccessful. Please try again.', 'Error updating'));
      });
  };
}

const ABOUT_CONTENT = `# About TalentMAP

TalentMAP is a multi-year, department-wide talent management platform expanding the Foreign Service (FS) bidding and assignments concept to become a much greater strategy for managing our FS talent worldwide to:

* Manage our assignment processes
* Align our people to positions
* Project our workforce planning of the future

## Our Mission

HR/CDA supports the Department of State (DOS) mission by transparently and equitably assigning employees with the right skills to the right positions at the right time.

## TalentMAP Progress is Ongoing

For more information, visit [HR Systems Online](http://www.state.gov).

## How TalentMAP Works
The Bureau of Human Resources Executive Office (HR/EX) in partnership with the Bureau of Human Resources Office of Career Development and Assignments (HR/CDA) are developing a new application, TalentMAP, aimed at replacing the Foreign Service Bidding (FSBid) application. TalentMAP will serve as a unified source of information regarding available FS positions and post data for Foreign Service employees. HR/EX aims to ensure TalentMAP becomes a secure, reliable, and central system that will support the HR/CDA business processes.

TalentMAP will maintain the existing business functionality and meet evolving requirements while leveraging best practices in software development. As a result of TalentMAP, DOS will be able to better carry out its diplomatic mission by improving the matching process, reducing the amount of time it takes for FS employees to find jobs that align with their skills, and improving the workflow for administrative staff to manage FS employees.


## Help Build TalentMAP
The first phase of the TalentMAP release will be optional for bidders offering a search and view-only option. The DOS employee will have the ability to search for potential positions, research posts, and add positions as “favorites.” Favorited positions will organize the bidder in order to make a determination on which positions they intend to bid.

During the first phase, DOS employees will need to return to FSBid to submit their bid(s) and continue the assignment process. FS bidders can choose to pilot the TalentMAP application to better understand the functionality and have the opportunity to provide feedback.
`;
