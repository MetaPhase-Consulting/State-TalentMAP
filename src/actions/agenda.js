import { downloadFromResponse, formatDate } from 'utilities';
import api from '../api';

// eslint-disable-next-line import/prefer-default-export
export function agendaItemHistoryExport(cdo, isSort = 'Position_Title') {
  console.log('agendatitemhisty action');
  return api()
    .get(`/fsbid/agenda_items/export/?ordering=${isSort}`)
    .then((response) => {
      downloadFromResponse(response, `Agenda_Item_History_${formatDate(new Date().getTime(), 'YYYY_M_D_Hms')}`);
    });
}

