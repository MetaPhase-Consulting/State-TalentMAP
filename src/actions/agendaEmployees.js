
import { batch } from 'react-redux';
import { get, identity, isArray, isString, pickBy } from 'lodash';
import { stringify } from 'query-string';
import { CancelToken } from 'axios';
import { downloadFromResponse, formatDate } from 'utilities';
import Q from 'q';
import api from '../api';

let cancelAgendaEmployees;

export function agendaEmployeesFetchDataErrored(bool) {
  return {
    type: 'AGENDA_EMPLOYEES_FETCH_HAS_ERRORED',
    hasErrored: bool,
  };
}

export function agendaEmployeesFetchDataLoading(bool) {
  return {
    type: 'AGENDA_EMPLOYEES_FETCH_IS_LOADING',
    isLoading: bool,
  };
}

export function agendaEmployeesFetchDataSuccess(results) {
  return {
    type: 'AGENDA_EMPLOYEES_FETCH_SUCCESS',
    results,
  };
}

export function agendaEmployeesFiltersFetchDataErrored(bool) {
  return {
    type: 'AGENDA_EMPLOYEES_FILTERS_FETCH_HAS_ERRORED',
    hasErrored: bool,
  };
}

export function agendaEmployeesFiltersFetchDataLoading(bool) {
  return {
    type: 'AGENDA_EMPLOYEES_FILTERS_FETCH_IS_LOADING',
    isLoading: bool,
  };
}

export function agendaEmployeesFiltersFetchDataSuccess(results) {
  return {
    type: 'AGENDA_EMPLOYEES_FILTERS_FETCH_SUCCESS',
    results,
  };
}

const convertQueryToString = query => {
  let q = pickBy(query, identity);
  Object.keys(q).forEach(queryk => {
    if (isArray(q[queryk])) { q[queryk] = q[queryk].join(); }
    if (isString(q[queryk]) && !q[queryk]) {
      q[queryk] = undefined;
    }
  });
  q = stringify(q);
  return q;
};

export function agendaItemHistoryExport(query = {}) {
  const q = convertQueryToString(query);
  const endpoint = '/fsbid/agenda_employees/export/';
  const ep = `${endpoint}?${q}`;
  return api()
    .get(ep)
    .then((response) => {
      downloadFromResponse(response, `Agenda_Item_Employees_${formatDate(new Date().getTime(), 'YYYY_M_D_Hms')}`);
    });
}

export function agendaEmployeesFetchData(query = {}) {
  return (dispatch) => {
    if (cancelAgendaEmployees) { cancelAgendaEmployees('cancel'); }
    batch(() => {
      dispatch(agendaEmployeesFetchDataLoading(true));
      dispatch(agendaEmployeesFetchDataErrored(false));
    });
    const q = convertQueryToString(query);
    const endpoint = '/fsbid/agenda_employees/';
    const ep = `${endpoint}?${q}`;
    api().get(ep, {
      cancelToken: new CancelToken((c) => {
        cancelAgendaEmployees = c;
      }),
    })
      .then(({ data }) => {
        batch(() => {
          dispatch(agendaEmployeesFetchDataSuccess(data));
          dispatch(agendaEmployeesFetchDataErrored(false));
          dispatch(agendaEmployeesFetchDataLoading(false));
        });
      })
      .catch((err) => {
        if (get(err, 'message') === 'cancel') {
          batch(() => {
            dispatch(agendaEmployeesFetchDataErrored(false));
            dispatch(agendaEmployeesFetchDataLoading(true));
          });
        } else {
          batch(() => {
            dispatch(agendaEmployeesFetchDataSuccess([]));
            dispatch(agendaEmployeesFetchDataErrored(true));
            dispatch(agendaEmployeesFetchDataLoading(false));
          });
        }
      });
  };
}

export function agendaEmployeesSelectionsSaveSuccess(result) {
  return {
    type: 'AGENDA_EMPLOYEES_SELECTIONS_SAVE_SUCCESS',
    result,
  };
}

export function saveAgendaEmployeesSelections(queryObject) {
  return (dispatch) => dispatch(agendaEmployeesSelectionsSaveSuccess(queryObject));
}

export function agendaEmployeesFiltersFetchData() {
  return (dispatch) => {
    batch(() => {
      dispatch(agendaEmployeesFiltersFetchDataLoading(true));
      dispatch(agendaEmployeesFiltersFetchDataErrored(false));
    });
    const ep = [
      '/fsbid/agenda_employees/reference/current-bureaus/',
      '/fsbid/agenda_employees/reference/handshake-bureaus/',
      '/fsbid/agenda_employees/reference/current-organizations/',
      '/fsbid/agenda_employees/reference/handshake-organizations/',
    ];
    const queryProms = ep.map(url =>
      api().get(url)
        .then((r) => r)
        .catch((e) => e),
    );
    Q.allSettled(queryProms)
      .then((results) => {
        const successCount = results.filter(r => r.state === 'fulfilled' && r.value).length || 0;
        const queryPromsLen = queryProms.length || 0;
        const countDiff = queryPromsLen - successCount;
        if (countDiff > 0) {
          batch(() => {
            dispatch(agendaEmployeesFiltersFetchDataErrored(true));
            dispatch(agendaEmployeesFiltersFetchDataLoading(false));
          });
        } else {
          const currentBureaus = get(results, '[0].value.data', []);
          const handshakeBureaus = get(results, '[1].value.data', []);
          const currentOrganizations = get(results, '[2].value.data', []);
          const handshakeOrganizations = get(results, '[3].value.data', []);
          const filters = {
            currentBureaus, handshakeBureaus, currentOrganizations, handshakeOrganizations,
          };
          batch(() => {
            dispatch(agendaEmployeesFiltersFetchDataSuccess(filters));
            dispatch(agendaEmployeesFiltersFetchDataErrored(false));
            dispatch(agendaEmployeesFiltersFetchDataLoading(false));
          });
        }
      })
      .catch(() => {
        batch(() => {
          dispatch(agendaEmployeesFiltersFetchDataErrored(true));
          dispatch(agendaEmployeesFiltersFetchDataLoading(false));
        });
      });
  };
}
