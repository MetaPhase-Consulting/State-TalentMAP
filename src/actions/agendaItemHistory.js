import { get } from 'lodash';
import { downloadFromResponse, formatDate } from 'utilities';
import { CancelToken } from 'axios';
import api from '../api';

let cancel;

export function aihHasErrored(bool) {
  return {
    type: 'AIH_HAS_ERRORED',
    hasErrored: bool,
  };
}
export function aihIsLoading(bool) {
  return {
    type: 'AIH_IS_LOADING',
    isLoading: bool,
  };
}
export function aihFetchDataSuccess(data) {
  return {
    type: 'AIH_FETCH_DATA_SUCCESS',
    data,
  };
}

export function agendaItemHistoryExport(perdet = '', ordering = '', client = '') {
  return api()
    .get(`/fsbid/agenda/agenda_items/export/?perdet=${perdet}&ordering=${ordering}&client=${client}`)
    .then((response) => {
      downloadFromResponse(response, `Agenda_Item_History_${client}_${formatDate(new Date().getTime(), 'YYYY_M_D_Hms')}`);
    });
}

export function aihFetchData(perdet = '', ordering = '') {
  return (dispatch) => {
    if (cancel) { cancel('cancel'); }
    if (!perdet) {
      dispatch(aihHasErrored(true));
      dispatch(aihIsLoading(false));
    } else {
      dispatch(aihHasErrored(false));
      dispatch(aihIsLoading(true));
      api()
        .get(`/fsbid/agenda/agenda_items/?perdet=${perdet}&ordering=${ordering}`, { cancelToken: new CancelToken((c) => {
          cancel = c;
        }) })
        .then(({ data }) => data || [])
        .then((data$) => {
          dispatch(aihFetchDataSuccess(data$));
          dispatch(aihHasErrored(false));
          dispatch(aihIsLoading(false));
        })
        .catch((err) => {
          if (get(err, 'message') === 'cancel') {
            dispatch(aihHasErrored(false));
            dispatch(aihIsLoading(true));
          } else {
            dispatch(aihHasErrored(true));
            dispatch(aihIsLoading(false));
          }
        });
    }
  };
}
