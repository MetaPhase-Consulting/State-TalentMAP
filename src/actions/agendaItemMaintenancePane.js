import { get } from 'lodash';
import { CancelToken } from 'axios';
import { batch } from 'react-redux';
import { UPDATE_AGENDA_ITEM_ERROR,
  UPDATE_AGENDA_ITEM_ERROR_TITLE, UPDATE_AGENDA_ITEM_SUCCESS,
  UPDATE_AGENDA_ITEM_SUCCESS_TITLE,
} from 'Constants/SystemMessages';
import { toastError, toastSuccess } from './toast';
import api from '../api';

let cancel;

export function aiCreateErrored(bool) {
  return {
    type: 'AI_CREATE_HAS_ERRORED',
    hasErrored: bool,
  };
}
export function aiCreateLoading(bool) {
  return {
    type: 'AI_CREATE_IS_LOADING',
    isLoading: bool,
  };
}
export function aiCreateSuccess(data) {
  return {
    type: 'AI_CREATE_SUCCESS',
    data,
  };
}

export function aiCreate(panel, legs, perSeqNum, ef) {
  // eslint-disable-next-line no-console
  console.log('ðŸ‘¾ current: ', {
    perSeqNum,
    ...panel,
    ...ef,
    agendaLegs: legs,
  });

  return (dispatch) => {
    if (cancel) { cancel('cancel'); }
    dispatch(aiCreateErrored(false));
    dispatch(aiCreateLoading(true));
    api()
      .post('/fsbid/agenda/agenda_item/', {
        perSeqNum,
        ...panel,
        ...ef,
        agendaLegs: legs,
      }, {
        cancelToken: new CancelToken((c) => {
          cancel = c;
        }),
      })
      .then(({ data }) => {
        batch(() => {
          dispatch(aiCreateErrored(false));
          dispatch(aiCreateSuccess(data || []));
          dispatch(toastSuccess(UPDATE_AGENDA_ITEM_SUCCESS, UPDATE_AGENDA_ITEM_SUCCESS_TITLE));
          dispatch(aiCreateLoading(true));
        });
      })
      .catch((err) => {
        if (get(err, 'message') === 'cancel') {
          dispatch(aiCreateErrored(false));
          dispatch(aiCreateLoading(true));
        } else {
          dispatch(toastError(UPDATE_AGENDA_ITEM_ERROR, UPDATE_AGENDA_ITEM_ERROR_TITLE));
          dispatch(aiCreateErrored(true));
          dispatch(aiCreateLoading(false));
        }
      });
  };
}
