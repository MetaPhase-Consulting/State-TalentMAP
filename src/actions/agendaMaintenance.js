import { batch } from 'react-redux';
import { get } from 'lodash';
import api from '../api';

export function agendaItemSaveErrored(bool) {
  return {
    type: 'AGENDA_ITEM_SAVE_ERRORED',
    hasErrored: bool,
  };
}

export function agendaItemSaveLoading(bool) {
  return {
    type: 'AGENDA_ITEM_SAVE_LOADING',
    isLoading: bool,
  };
}

export function agendaItemSaveSuccess(results) {
  return {
    type: 'AGENDA_ITEM_SAVE_SUCCESS',
    results,
  };
}

export function saveAgendaItem(ai_data = {}) {
  // eslint-disable-next-line no-console
  console.log('current: ai_data', ai_data);
  return (dispatch) => {
    batch(() => {
      dispatch(agendaItemSaveLoading(true));
      dispatch(agendaItemSaveErrored(false));
    });
    api().post('/fsbid/save_ai/')
      .then(({ data }) => {
        batch(() => {
          dispatch(agendaItemSaveSuccess(data));
          dispatch(agendaItemSaveErrored(false));
          dispatch(agendaItemSaveLoading(false));
        });
      })
      .catch((err) => {
        if (get(err, 'message') === 'cancel') {
          batch(() => {
            dispatch(agendaItemSaveErrored(false));
            dispatch(agendaItemSaveLoading(true));
          });
        } else {
          batch(() => {
            dispatch(agendaItemSaveSuccess([]));
            dispatch(agendaItemSaveErrored(true));
            dispatch(agendaItemSaveLoading(false));
          });
        }
      });
  };
}
