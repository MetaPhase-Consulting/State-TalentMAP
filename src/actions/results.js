import { CancelToken } from 'axios';
import queryString from 'query-string';
import { get } from 'lodash';
import api from '../api';
import { checkFlag } from '../flags';

const getUseAP = () => checkFlag('flags.available_positions');

let cancel;
let cancelSimilar;

export function resultsHasErrored(bool) {
  return {
    type: 'RESULTS_HAS_ERRORED',
    hasErrored: bool,
  };
}
export function resultsIsLoading(bool) {
  return {
    type: 'RESULTS_IS_LOADING',
    isLoading: bool,
  };
}
export function resultsFetchDataSuccess(results) {
  return {
    type: 'RESULTS_FETCH_DATA_SUCCESS',
    results,
  };
}

export function resultsSimilarPositionsHasErrored(bool) {
  return {
    type: 'RESULTS_SIMILAR_POSITIONS_HAS_ERRORED',
    hasErrored: bool,
  };
}
export function resultsSimilarPositionsIsLoading(bool) {
  return {
    type: 'RESULTS_SIMILAR_POSITIONS_IS_LOADING',
    isLoading: bool,
  };
}
export function resultsSimilarPositionsFetchDataSuccess(results) {
  return {
    type: 'RESULTS_SIMILAR_POSITIONS_FETCH_DATA_SUCCESS',
    results,
  };
}

export function resultsFetchSimilarPositions(id) {
  return (dispatch) => {
    if (cancelSimilar) { cancelSimilar(); }
    const useAP = getUseAP();
    const prefix = useAP ? '/fsbid/available_positions' : '/cycleposition';

    dispatch(resultsSimilarPositionsIsLoading(true));
    api().get(`${prefix}/${id}/similar/?limit=3`, {
      cancelToken: new CancelToken((c) => {
        cancelSimilar = c;
      }),
    })
      .then(response => response.data)
      .then((results) => {
        dispatch(resultsSimilarPositionsFetchDataSuccess(results));
        dispatch(resultsSimilarPositionsHasErrored(false));
        dispatch(resultsSimilarPositionsIsLoading(false));
      })
      .catch(() => {
        dispatch(resultsSimilarPositionsFetchDataSuccess({}));
        dispatch(resultsSimilarPositionsHasErrored(true));
        dispatch(resultsSimilarPositionsIsLoading(false));
      });
  };
}

export function downloadPositionData(query, isPV) {
  const prefix = `fsbid/${!isPV ? '/fsbid/available_positions' : '/fsbid/projected_vacancies'}/export`;
  return api()
  .get(`${prefix}/?${query}`, {
    cancelToken: new CancelToken((c) => { cancel = c; }),
    responseType: 'stream',
  })
  .then((response) => {
    const cd = get(response, 'headers.content-disposition');
    const filename = cd.replace('attachment; filename=', '') || 'TalentMap_search_export';
    const url = window.URL.createObjectURL(new Blob([response.data]));
    const a = document.createElement('a');
    a.href = url;
    a.setAttribute('download', filename);
    document.body.appendChild(a);
    a.click();
  });
}

export function fetchResultData(query) {
  const useAP = getUseAP();

  let prefix = useAP ? '/fsbid/available_positions' : '/cycleposition';
  const parsed = queryString.parse(query);
  const isPV = parsed.projectedVacancy;

  if (isPV) {
    prefix = '/fsbid/projected_vacancies';
    delete parsed.projectedVacancy;
  }

  const query$ = queryString.stringify(parsed);

  return api()
  .get(`${prefix}/?${query$}`, {
    cancelToken: new CancelToken((c) => { cancel = c; }),
  })
  .then((response) => {
    if (isPV) {
      return {
        ...response.data,
        isProjectedVacancy: true,
      };
    }
    return response.data;
  });
}

export function resultsFetchData(query) {
  return (dispatch) => {
    if (cancel) { cancel('cancel'); dispatch(resultsIsLoading(true)); }
    dispatch(resultsIsLoading(true));
    dispatch(resultsHasErrored(false));
    fetchResultData(query)
      .then((results) => {
        dispatch(resultsFetchDataSuccess(results));
        dispatch(resultsHasErrored(false));
        dispatch(resultsIsLoading(false));
      })
      .catch((err) => {
        if (get(err, 'message') === 'cancel') {
          dispatch(resultsHasErrored(false));
          dispatch(resultsIsLoading(true));
        } else {
          dispatch(resultsFetchDataSuccess({ results: [] }));
          dispatch(resultsHasErrored(true));
          dispatch(resultsIsLoading(false));
        }
      });
  };
}
