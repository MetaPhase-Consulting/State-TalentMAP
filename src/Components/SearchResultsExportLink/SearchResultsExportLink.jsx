import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { get, mapValues } from 'lodash';
import queryString from 'query-string';
import flatten from 'flat';
import { CSVLink } from '../CSV';
import { POSITION_SEARCH_SORTS } from '../../Constants/Sort';
import { fetchResultData } from '../../actions/results';
import { formatDate, getFormattedNumCSV, spliceStringForCSV } from '../../utilities';
import ExportButton from '../ExportButton';

// Mapping columns to data fields
// Use custom delimiter of flattened data
const HEADERS = [
  { label: 'Position', key: 'position__title' },
  { label: 'Position number', key: 'position_number' },
  { label: 'Skill', key: 'position__skill' },
  { label: 'Grade', key: 'grade' },
  { label: 'Bureau', key: 'position__bureau' },
  { label: 'Post city', key: 'position__post__location__city' },
  { label: 'Post country', key: 'position__post__location__country' },
  { label: 'Tour of duty', key: 'position__post__tour_of_duty' },
  { label: 'Language', key: 'position__languages__0__representation' },
  { label: 'Post differential', key: 'position__post__differential_rate' },
  { label: 'Danger pay', key: 'position__post__danger_pay' },
  { label: 'TED', key: 'estimated_end_date' },
  { label: 'Incumbent', key: 'position__current_assignment__user' },
  { label: 'Bid Cycle/Season', key: 'bidcycle__name' },
  { label: 'Posted date', key: 'posted_date' },
  { label: 'Status code', key: 'status_code' },
];

// Processes results before sending to the download component to allow for custom formatting.
// Flatten data with custom delimiter.
export const processData = data =>
  data.map((entry) => {
    const entry$ = flatten({ ...entry }, { delimiter: '__' });
    const endDate = get(entry$, 'ted');
    const postedDate = get(entry$, 'posted_date');
    const formattedEndDate = endDate ? formatDate(endDate) : null;
    const formattedPostedDate = postedDate ? formatDate(postedDate) : null;
    const val = {
      ...mapValues(entry$, x => !x ? '' : x), // eslint-disable-line no-confusing-arrow
      position_number: getFormattedNumCSV(entry.position.position_number),
      grade: getFormattedNumCSV(get(entry, 'position.grade')),
      estimated_end_date: formattedEndDate,
      posted_date: formattedPostedDate,
    };
    return val;
  });


class SearchResultsExportLink extends Component {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
    this.state = {
      isLoading: false,
      data: '',
      query: { value: window.location.search.replace('?', '') || '' },
    };
  }

  componentWillReceiveProps() {
    const query = window.location.search.replace('?', '') || '';
    if (this.state.query.value !== query) {
      this.setState({ query: { value: query } });
    }
  }

  onClick() {
    const { isLoading } = this.state;
    if (!isLoading) {
      // reset the state to support multiple clicks
      this.setState({ data: '', isLoading: true });
      const query = {
        ordering: POSITION_SEARCH_SORTS.defaultSort,
        ...queryString.parse(this.state.query.value),
        limit: this.props.count,
        page: 1,
      };
      fetchResultData(queryString.stringify(query))
      .then((results) => {
        const data = processData(results.results);
        this.setState({ data, isLoading: false }, () => {
          // click the CSVLink component to trigger the CSV download
          // This is needed for the download to work in Edge.
          if (this.csvLink) { this.csvLink.link.click(); }
        });
      })
      .catch(() => {
        this.setState({ isLoading: false });
      });
    }
  }

  render() {
    const { data, isLoading } = this.state;
    return (
      <div className="export-button-container">
        <ExportButton onClick={this.onClick} isLoading={isLoading} />
        <CSVLink
          tabIndex="-1"
          transform={spliceStringForCSV}
          ref={(x) => { this.csvLink = x; }}
          target="_blank"
          filename={this.props.filename}
          data={data}
          headers={HEADERS}
          uFEFF={false}
        />
      </div>
    );
  }
}

SearchResultsExportLink.propTypes = {
  count: PropTypes.number,
  filename: PropTypes.string,
};

SearchResultsExportLink.defaultProps = {
  count: 0,
  filename: 'TalentMap_search_export.csv',
};

export default SearchResultsExportLink;
