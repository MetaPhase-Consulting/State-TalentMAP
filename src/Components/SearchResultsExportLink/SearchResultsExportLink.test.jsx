import { shallow } from 'enzyme';
import React from 'react';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import SearchResultsExportLink, { processData } from './SearchResultsExportLink';
import * as actions from '../../actions/results';

describe('SearchResultsExportLink', () => {
  let fetchResultDataStub;

  beforeEach(() => {
    fetchResultDataStub = sinon.stub(actions, 'fetchResultData').returns(Promise.resolve({ results: [] }));
  });

  afterEach(() => {
    actions.fetchResultData.restore();
  });

  it('gets data when button is clicked', (done) => {
    const wrapper = shallow(<SearchResultsExportLink />);
    expect(wrapper.state().data).toBeFalsy();
    wrapper.find('ExportButton').props().onClick();
    expect(fetchResultDataStub.calledOnce).toBe(true);
    const f = () => setTimeout(() => {
      expect(wrapper.instance().state.data).toEqual([]);
      done();
    }, 100);
    f();
  });

  it('shows link component when state has data', () => {
    const wrapper = shallow(<SearchResultsExportLink />);
    wrapper.setState({ data: 'test' });
    wrapper.update();
    expect(wrapper.find('CSVLink').prop('data')).toBeTruthy();
  });

  it('sets state on componentWillReceiveProps()', () => {
    const wrapper = shallow(<SearchResultsExportLink />);
    window.history.pushState({}, '', '?stuff=1&things=5');
    wrapper.instance().componentWillReceiveProps();
    expect(wrapper.instance().state.query.value).toBe('stuff=1&things=5');
  });

  it('processes data correctly', () => {
    const data = [{ a: 1, b: 2, position: { current_assignment: { estimated_end_date: '2019-01-08T00:00:00Z' } } }];
    const output = processData(data);
    expect(output[0]).toMatchObject(
      { a: 1,
        b: 2,
        position: {
          current_assignment: { estimated_end_date: '2019-01-08T00:00:00Z' },
        },
        estimated_end_date: '01/07/2019',
        grade: '',
        position_number: '',
      },
    );
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<SearchResultsExportLink />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
