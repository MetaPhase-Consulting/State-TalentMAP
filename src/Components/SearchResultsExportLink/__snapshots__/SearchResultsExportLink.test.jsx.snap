// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsExportLink matches snapshot 1`] = `
<div
  className="export-button-container"
>
  <ExportButton
    className=""
    isLoading={false}
    onClick={[Function]}
    text="Export"
  />
  <CSVLink
    asyncOnClick={false}
    data=""
    filename="TalentMap_search_export.csv"
    headers={
      Array [
        Object {
          "key": "title",
          "label": "Position",
        },
        Object {
          "key": "position_number",
          "label": "Position number",
        },
        Object {
          "key": "skill",
          "label": "Skill",
        },
        Object {
          "key": "grade",
          "label": "Grade",
        },
        Object {
          "key": "bureau",
          "label": "Bureau",
        },
        Object {
          "key": "post.location.city",
          "label": "Post city",
        },
        Object {
          "key": "post.location.country",
          "label": "Post country",
        },
        Object {
          "key": "post.tour_of_duty",
          "label": "Tour of duty",
        },
        Object {
          "key": "languages[0].representation",
          "label": "Language",
        },
        Object {
          "key": "post.differential_rate",
          "label": "Post differential",
        },
        Object {
          "key": "post.danger_pay",
          "label": "Danger pay",
        },
        Object {
          "key": "estimated_end_date",
          "label": "TED",
        },
        Object {
          "key": "current_assignment.user",
          "label": "Incumbent",
        },
      ]
    }
    separator=","
    tabIndex="-1"
    target="_blank"
    transform={[Function]}
    uFEFF={false}
  />
</div>
`;
